graylog2-server/src/main/java/org/graylog2/shared/bindings/GenericBindings.java

```diff
@@ -90,6 +94,12 @@ protected void configure() {
                 .setDefault()
                 .to(DefaultFailureHandlingConfiguration.class);
 
+        final Multibinder<IndexTemplateProvider> indexTemplateProviderMultibinder = Multibinder.newSetBinder(binder(), IndexTemplateProvider.class);
+
+        indexTemplateProviderMultibinder.addBinding().to(MessageIndexTemplateProvider.class);
+        indexTemplateProviderMultibinder.addBinding().to(EventIndexTemplateProvider.class);
+        indexTemplateProviderMultibinder.addBinding().to(GimIndexTemplateProvider.class);
+
```

Comment:
We usually use a map binder and use the type string as key. Using a set binder and then find factories by iterating over the set and comparing the type works as well, of course. Any specific reason why you didn't choose a map binder?

Apart from the easier lookup, another advantage of the map binder is, that trying to bind the same key twice results in a binding error on server startup. With the set binder we have to check that there is only a single implementation for the same key at runtime and throw an IllegalStateException.

Reply:
https://github.com/Graylog2/graylog2-server/pull/11290/commits/abcd797052c00b60b2e09380fa5085ed706e14f8